### PR TITLE
Add support for dynamic positional/keyword inputs during training for ORTModule

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
Because ORTModule uses tracing for exporting `torch.nn.Module` models to ONNX, the graph has static input signature. That means the number of inputs and their types during the first feed forward will be expected to be respected throughout training. 

Many models, on the other hand, changes the number of input arguments, or their type/shape, during the course of training.

One class of mutating inputs are related to variable positional arguments (*args): 
Suppose the first call to `model(input1, *args)` in the training script is `model(torch.randn(2,3,dtype=torch.float32))`. That would result in a graph being created with a single input, with type float and shape (2,3). However, during training input could change to `model(torch.randn(2,2))` or `model(torch.randn(2,3, dtype=torch.half) or even `model(torch.randn(2,3), torch.randn(1,2,3,dtype=torch.half))`

A second class of mutating inpust are related to variable keyword arguments (**kwargs):
Suppose the first call to `model(input1=None,input2=None,input3=None)` in the training script is `model(input1=torch.randn(2,3,dtype=torch.float32))`. That would result in a graph being created with a single input called input1, with type float and shape (2,3). However, during training input could change to `model(input1=torch.randn(2,2))` or `model(input1=torch.randn(2,3, dtype=torch.half) or even `model(input1=torch.randn(2,3), input2=torch.randn(1,2,3,dtype=torch.half),input3=torch.ones(3,4))`

ORT would either failing executing such graph as the expected input does not match the actual input received during training or it would execute the graph ignoring the extra input.

Today, ORTModule supports training when the model takes completely static inputs or dynamic `*args`/`**kwrags` as input, but the number of arguments, their type must match in all batches, from the first iteration to the last. Shape-wise, the value of each dimension can be different during training, but the number of dimensions must match the first forward call!

This PR add support for scenarios in which there might be different number of arguments for both positional and keyword inputs or their types and shapes may change from one forward call to next, as exemplified above

The solution is basically building and keeping a schema for all the *args and **kwargs during model export and compare to the schema of the new incoming data. If a change is detected, model is re-exported. The schema only contains name (for dicts), type and number of dimensions (as opposed to shape). We do not consider the shape itself as we support dynamic axis and if the number of dimensions are same, we dont care about their actual values

Tests include scenarios with both varying positional and keyword arguments in both training and evaluation mode

### CODE REVIEW

File: orttraining/orttraining/python/training/_ortmodule_io.py [hunk 1]
```diff
@@ -5,6 +5,55 @@
...
 import warnings
 
+
+class _InputInfo(object):
+    def __init__(self, names, shape, require_grad_names = [], dynamic_axes = {}, schema = []):
+        self.names = names
+        self.shape = shape
+        self.require_grad_names = require_grad_names
+        self.dynamic_axes = dynamic_axes
+        self.schema = schema
```
Comment (mrry) [inline orttraining/orttraining/python/training/_ortmodule_io.py:15]: I'm surprised our linter doesn't complain about this, but it's generally a good idea to use `None` as the defaults for potentially-mutable default args, just in case someone comes along and modifies the default and creates a hard-to-debug issue:

```suggestion
    def __init__(self, names, shape, require_grad_names=None, dynamic_axes=None, schema=None):
        self.names = names
        self.shape = shape
        self.require_grad_names = require_grad_names if require_grad_names else []
        self.dynamic_axes = dynamic_axes if dynamic_axes else {}
        self.schema = schema if schema else []
```

Alternatively, since there's only one callsite, and it passes all of the arguments, you could make all of them mandatory.

File: orttraining/orttraining/python/training/_ortmodule_io.py [hunk 2]
```diff
@@ -5,6 +5,55 @@
...
+
+    def id(self) -> int:
+        return hash(repr(self.schema))
```
Comment (mrry) [inline orttraining/orttraining/python/training/_ortmodule_io.py:18]: Is this used anywhere? I don't think it's overriding the behavior of `id()`, if that's what you're intending....

Perhaps you want to override `__hash__()` and/or `__eq__()`?

File: orttraining/orttraining/python/training/ortmodule.py [hunk 3]
```diff
@@ -236,14 +250,14 @@ def backward(ctx, *grad_outputs):
...
-                    num_user_input_grads = len(self._input_names_require_grad)
+                    num_user_input_grads = len(self._input_info.require_grad_names)
 
                     results = []
                     for input_name in self._onnx_graphs_info.user_input_names:
                         try:
                             # Append to the results the backward output for each input that required grad
                             results.append(_ortvalue_to_torch_tensor(
-                                backward_outputs[self._input_names_require_grad.index(input_name)]))
+                                backward_outputs[self._input_info.require_grad_names.index(input_name)]))
```
Comment (mrry) [inline orttraining/orttraining/python/training/ortmodule.py:260]: As part of the refactoring in this PR, can we replace (or supplement) the `require_grad_names` list with a set, so that this loop isn't O(n^2) in the number of tensors that require a gradient?

**Review by mrry — CHANGES_REQUESTED** [review]

File: orttraining/orttraining/python/training/_ortmodule_io.py (hunk 2 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/_ortmodule_io.py:18]: This is a leftover. Removing it!

File: orttraining/orttraining/python/training/_ortmodule_io.py [hunk 4]
```diff
@@ -18,9 +67,49 @@ def deepcopy_model_input(*inputs, **kwargs):
...
+        if self.shape_dims is not None:
+            if result[-1] != '(':
+                result += ', '
+            result += f'shape_dims={self.shape_dims}'
+        result += ')'
+        return result
+
+    def __eq__(self, other):
+        if not isinstance(other, _TensorStub):
+            return NotImplemented
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/_ortmodule_io.py:101]: Should this raise?

File: orttraining/orttraining/python/training/_ortmodule_io.py [hunk 5]
```diff
@@ -18,9 +67,49 @@ def deepcopy_model_input(*inputs, **kwargs):
...
+        elif not other:
+            return False
+        elif self.name != self.name != other.name:
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/_ortmodule_io.py:104]: unintentional: All checks have been duplicated :
```self.name != self.name !=other.name```.

File: orttraining/orttraining/python/training/ortmodule.py (hunk 3 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/ortmodule.py]: Sure, this change came from the original YieldOp merge, but I can take a look into it

File: orttraining/orttraining/python/training/_ortmodule_io.py [hunk 6]
```diff
@@ -5,6 +5,52 @@
...
+    result = []
+    for input_idx, name in enumerate(user_input_names):
+        inp = None
+        if input_idx < len(non_none_inputs):
+            inp = non_none_inputs[input_idx]
+        elif name in kwargs and kwargs[name] is not None:
+            inp = kwargs[name]
+        elif input_idx >= len(non_none_inputs):
+            # Registered buffers are translated to user_input+initializer in ONNX
+            # TODO: Check what happens when the number of inputs change form one call to the next
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/_ortmodule_io.py:39]: Does this TODO need to be removed?

File: orttraining/orttraining/python/training/_ortmodule_io.py [hunk 7]
```diff
@@ -5,6 +5,52 @@
...
+            buffer_name, inp = next(named_buffers_iter)
+            assert buffer_name == name, f'Input name {name} expected, but {buffer_name} found!'
+
+        if inp is not None:
+            result.append(inp)
+        else:
+            # TODO: Re-export ONNX if any input from _onnx_graphs_info.user_input_names is None.
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/_ortmodule_io.py:46]: I think we probably can remove this TODO and exception raising (change it to an assert) also?
Basically, when we get to this stage, there should be no None input, right?

File: orttraining/orttraining/python/training/_ortmodule_io.py [hunk 8]
```diff
@@ -18,9 +64,49 @@ def deepcopy_model_input(*inputs, **kwargs):
...
 
-class _TensorStub:
-    # Stub for a torch.Tensor value to be used to formulate the output schema
-    pass
+
+class _TensorStub(object):
+    '''Tensor stub class used to represent model's input or output'''
+
+    def __init__(self, name=None, dtype=None, shape=None, shape_dims=None):
+        self.name = name
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/_ortmodule_io.py:72]: Shall we make all these attributes internal using the _ prefix?

File: orttraining/orttraining/python/training/_ortmodule_io.py [hunk 9]
```diff
@@ -18,9 +64,49 @@ def deepcopy_model_input(*inputs, **kwargs):
...
+        if self.shape_dims is not None:
+            if result[-1] != '(':
+                result += ', '
+            result += f'shape_dims={self.shape_dims}'
+        result += ')'
+        return result
+
+    def __eq__(self, other):
+        if not isinstance(other, _TensorStub):
+            raise NotImplemented
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/_ortmodule_io.py:98]: Can we also give a good error message here if we are throwing an exception?

File: orttraining/orttraining/python/training/_ortmodule_io.py [hunk 10]
```diff
@@ -64,33 +150,34 @@ def _replace_stub_with_tensor_value(user_output, outputs, output_idx):
...
-        return _TensorStub()
-
-    if isinstance(output, abc.Sequence):
-        sequence_type = type(output)
-        output = list(output)
-        for idx in range(len(output)):
-            output[idx] = _extract_output_schema(output[idx])
+    # Depth first traversal to iterate over the data to replace every tensor with a stub
+    elif isinstance(data, torch.Tensor):
+        return _TensorStub(None, str(data.dtype), None, len(data.size()))
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/_ortmodule_io.py:161]: We should probably explicitly set the attributes using assignment (instead of passing arguments positionally).

File: orttraining/orttraining/python/training/ortmodule.py [hunk 11]
```diff
@@ -112,38 +115,43 @@ def _forward(self, *inputs, **kwargs):
...
-                _ortmodule_io.parse_inputs_for_onnx_export(
-                    self._original_module_parameters, self._onnx_inference, *inputs, **kwargs)
+                        raise RuntimeError('A device must be specified by the model!')
+
+                self._onnx_inference = self._get_inference_graph(*inputs, **kwargs)
+                if self._save_onnx:
+                    onnx.save(self._onnx_inference, self._save_onnx_prefix + '_inference.onnx')
+                self._initialize_module_gradient_graph_builder()
+
+            input_info, input_info.schema = _io.parse_inputs_for_onnx_export(self._original_module_parameters,
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule.py:136]: Can we break this down into two separate lines? I find it much easier to read it that way (specially for longer sentences).

File: orttraining/orttraining/python/training/_ortmodule_io.py (hunk 8 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/_ortmodule_io.py:72]: This is already an internal data-class, never seen/used by users. Attributes must start with `_` when they are private and shouldn't be directly accessed

File: orttraining/orttraining/python/training/_ortmodule_io.py (hunk 7 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/_ortmodule_io.py:46]: Not entirely sure. I will remove the todo but keep the exception until we narrow all cases down

**Review by baijumeswani — DISMISSED** [review]
Looks good.

**Review by baijumeswani — APPROVED** [review]

File: orttraining/orttraining/python/training/ortmodule.py [hunk 12]
```diff
@@ -112,38 +115,44 @@ def _forward(self, *inputs, **kwargs):
...
+                self._onnx_inference = self._get_inference_graph(*inputs, **kwargs)
+                if self._save_onnx:
+                    onnx.save(self._onnx_inference, self._save_onnx_prefix + '_inference.onnx')
+                self._initialize_module_gradient_graph_builder()
+
+            input_info = _io.parse_inputs_for_onnx_export(self._original_module_parameters,
+                                                          self._onnx_inference,
+                                                          *inputs,
+                                                          **kwargs)
+            input_info.schema = schema
```
Comment (mrry) [inline orttraining/orttraining/python/training/ortmodule.py:140]: Is this line necessary? It looks like the same schema is extracted and assigned to this field inside `parse_inputs_for_onnx_export()`. (Could we reduce it to computing the schema once?)

File: orttraining/orttraining/python/training/ortmodule.py [hunk 13]
```diff
@@ -112,38 +115,44 @@ def _forward(self, *inputs, **kwargs):
...
+
+                self._onnx_inference = self._get_inference_graph(*inputs, **kwargs)
+                if self._save_onnx:
+                    onnx.save(self._onnx_inference, self._save_onnx_prefix + '_inference.onnx')
+                self._initialize_module_gradient_graph_builder()
+
+            input_info = _io.parse_inputs_for_onnx_export(self._original_module_parameters,
+                                                          self._onnx_inference,
+                                                          *inputs,
+                                                          **kwargs)
```
Comment (mrry) [inline orttraining/orttraining/python/training/ortmodule.py:139]: Instead of passing `*inputs, **kwargs`, could we directly pass `inputs, kwargs` (and remove the corresponding `*` and `**` in the `parse_inputs_for_onnx_export()` signature)? My intuition here is that this will avoid a bunch of data structure shallow copies and dynamic allocations in the call path.

File: orttraining/orttraining/python/training/ortmodule.py [hunk 14]
```diff
@@ -112,38 +115,44 @@ def _forward(self, *inputs, **kwargs):
...
+            input_info.schema = schema
+
+            initializer_names_to_train_set_user_model = {
+                name for name, param in self._flattened_output_module.named_parameters() if param.requires_grad}
             initializer_names_to_train_set_onnx_graph = set(self._onnx_graphs_info.initializer_names_to_train) \
```
Comment (mrry) [inline orttraining/orttraining/python/training/ortmodule.py:144]: Not in this PR, but... can we cache this set across calls to `forward()`?

File: orttraining/orttraining/python/training/ortmodule.py [hunk 15]
```diff
@@ -237,15 +251,18 @@ def backward(ctx, *grad_outputs):
                     backward_outputs = training_io_binding.get_outputs()
 
                     # Return input and initializer gradients
-                    num_user_input_grads = len(self._input_names_require_grad)
+                    num_user_input_grads = len(self._input_info.require_grad_names)
 
                     results = []
+                    require_grad_names_set = set(self._input_info.require_grad_names)
```
Comment (mrry) [inline orttraining/orttraining/python/training/ortmodule.py:257]: Suggestion: Compute this set once when the `InputInfo` is built.

File: orttraining/orttraining/python/training/ortmodule.py [hunk 16]
```diff
@@ -296,7 +318,7 @@ def backward(ctx, *grad_outputs):
         self._original_module = module
         # Get the module that flattens the output from the original module into a tuple
         self._flattened_output_module = \
-            _ortmodule_io.get_flattened_output_module(
+            _io.get_flattened_output_module(
                 self._original_module)
```
Comment (mrry) [inline orttraining/orttraining/python/training/ortmodule.py:322]:
```suggestion
        self._flattened_output_module = _io.get_flattened_output_module(self._original_module)
```

**Review by mrry — CHANGES_REQUESTED** [review]
Just a few more small suggestions: I've received some reports about performance overhead in the Python part of ORTModule, and I'd like to start chipping away at it with the parts that change in this PR.

(We should probably set up some more systematic microbenchmarks for the frontend to catch these issues.)

File: orttraining/orttraining/python/training/ortmodule.py (hunk 14 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/ortmodule.py:144]: We can create a PR for caching/optimizing some spots

File: orttraining/orttraining/python/training/ortmodule.py (hunk 15 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/ortmodule.py:257]: I will create a PR that caches this set along with the initializers_names_to_train

**Comment by thiagocrepaldi** [comment]
> Just a few more small suggestions: I've received some reports about performance overhead in the Python part of ORTModule, and I'd like to start chipping away at it with the parts that change in this PR.
> 
> (We should probably set up some more systematic microbenchmarks for the frontend to catch these issues.)

we certainly do!

**Review by baijumeswani — APPROVED** [review]

File: orttraining/orttraining/python/training/_ortmodule_utils.py [hunk 17]
```diff
@@ -5,58 +5,85 @@
...
 
-import onnxruntime
 from onnxruntime.capi.onnxruntime_inference_collection import OrtValue
 from onnxruntime.capi import _pybind_state as C
 
 import torch
 from torch.utils.dlpack import from_dlpack, to_dlpack
 from torch.utils.cpp_extension import load_inline
 
+from enum import IntEnum
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/_ortmodule_utils.py:15]: I think this should be a part of ```_ortmodule_logger.py```. Maybe the merge conflict resolution did not apply well here.

**Review by baijumeswani — APPROVED** [review]
